### DISCUSSION TITLE
`this` have different values in different scopes

### DISCUSSION THREAD

**Discussion by vdata1** [oven-sh/bun#13177] [General]
### What version of Bun is running?

1.1.20

### What platform is your computer?

Darwin 23.5.0 x86_64 i386

### What steps can reproduce the bug?

Hi, 

By considering the code snippet below, bun gives different values of `this` object in different scopes compared with node.js and deno.

I am wondering if it is not a bug, and why it is different. 
```javascript 
console.log(this); //binds to {} in bun and node, undefined in deno. 
(function () {
  console.log(this); //this binds to undefined in function block in deno and bun, global object in node.     
})();
```

AH

**Comment by KilianB** [comment]
Without digging into the details and checking it myself, bun by default initialises the project in strict mode in which the global this is undefined. Could that be a difference between your bun and node setup?

**Comment by vdata1** [comment]
Hi, @KilianB  thanks for the feedback, by looking at the code and the comments in, there is another raised question regarding deno and bun: since both implement `strict mode`  and initialise it by default, why do they behave differently at the global scope?, `Deno` gives `undefined` while `bun` gives `{}` like non `strict mode` `node`. Do they implement `strict mode` differently? 

I am glad to hear more from you. I really appreciate any help you can provide.

**Comment by codehz** [comment]
I think the top level this is another story...

from <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this>
> At the top level of a script, this refers to [globalThis](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) whether in strict mode or not. This is generally the same as the global object — for example, if the source is put inside an HTML [<script>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script) element and executed as a script, this === window.
> If the source is loaded as a [module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) (for HTML, this means adding type="module" to the <script> tag), this is always undefined at the top level.
> If the source is executed with [eval()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval), this is the same as the enclosing context for [direct eval](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#direct_and_indirect_eval), or globalThis (as if it's run in a separate global script) for indirect eval.
> Note that some source code, while looking like the global scope, is actually wrapped in a function when executed. For example, Node.js CommonJS modules are wrapped in a function and executed with the this value set to module.exports. [Event handler attributes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this#this_in_inline_event_handlers) are executed with this set to the element they are attached to.

so the problem here is, bun run code in module mode by default, but still follow the commonjs's  behavior, I guess it is for compatible
